src/react/hooks/use-presence.ts

```diff
@@ -0,0 +1,128 @@
+import { StatusParams } from '../status-params.js';
+import { useChatConnection } from './use-chat-connection.js';
+import { useRoom } from './use-room.js';
+
+/**
+ * The options for the {@link usePresence} hook.
+ */
+export interface UsePresenceParams extends StatusParams {
+  /**
+   * The data to enter the room with. Any JSON serializable data can be provided.
+   */
+  data?: PresenceData;
+}
+
+export interface UsePresenceResponse extends ChatStatusResponse {
+  /**
+   * A shortcut to the {@link Presence.update} method.
+   */
+  readonly update: Presence['update'];
+
+  /**
+   * A shortcut to the {@link Presence.leave} method.
+   */
+  readonly leave: Presence['leave'];
+
+  /**
+   * Indicates whether the current user is present in the room.
+   */
+  readonly isPresent: boolean;
+
+  /**
+   * Provides access to the underlying {@link Presence} instance of the room.
+   */
+  readonly presence: Presence;
+}
+
+/**
+ * A hook that provides access to the {@link Presence} instance in the room.
+ * It will use the instance belonging to the room in the nearest {@link RoomProvider} in the component tree.
+ * On calling, the hook will enter the room with the provided data and leave the room when the component unmounts.
+ * The {@link isPresent} flag will indicate when the user has become present in the room.
+ *
+ * @param params - Allows the registering of optional callbacks.
+ * @returns UsePresenceResponse - An object containing the {@link Presence} instance and methods to interact with it.
+ */
+export const usePresence = (params?: UsePresenceParams): UsePresenceResponse => {
+  const { currentStatus: connectionStatus, error: connectionError } = useChatConnection({
+    onStatusChange: params?.onConnectionStatusChange,
+  });
+  const { room, roomError, roomStatus } = useRoom({
+    onStatusChange: params?.onRoomStatusChange,
+  });
+
+  const [isPresent, setIsPresent] = useState(false);
+
+  // store the roomStatus in a ref to ensure the correct value is used in the effect cleanup
+  const roomStatusRef = useRef(roomStatus);
+
+  // we can't use the data param directly in a dependency array as it will cause an infinite loop
+  const dataRef = useRef(params?.data);
+  useEffect(() => {
+    dataRef.current = params?.data;
+  }, [params?.data]);
+
+  useEffect(() => {
+    // Update the ref when roomStatus changes
+    roomStatusRef.current = roomStatus;
+  }, [roomStatus]);
+
+  // enter the room when the hook is mounted
+  useEffect(() => {
+    // wait until the room is attached before attempting to enter
+    if (roomStatus !== RoomLifecycle.Attached) return;
+    void room.presence.enter(dataRef.current).then(() => {
+      setIsPresent(true);
+    })
+
+    return () => {
+      // ensure we are still in an attached state before attempting to leave;
+      // a presence.leave call will produce an exception otherwise.
+      if (roomStatusRef.current === RoomLifecycle.Attached) {
```

Comment:
We also need to check the connection state here, as an inactive connection will cause an exception:

https://github.com/ably/ably-js/blob/8ba8ec29f521b18ffd16bc4fcb33004c64a7d706/src/platform/react-hooks/src/hooks/usePresence.ts#L78